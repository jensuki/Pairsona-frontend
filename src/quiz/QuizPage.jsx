import { useState, useEffect, useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import Api from '../helpers/api';
import UserContext from '../context/UserContext';
import Loading from '../components/Loading';
import '../styles/QuizPage.css';

const QuizPage = () => {
    const { currUser, setCurrUser, loading } = useContext(UserContext);
    const [questions, setQuestions] = useState(null);
    const [answers, setAnswers] = useState({}); // track answers
    const [currentQuestionIndex, setCurrentQuestionIndex] = useState(0); // track current question
    const [error, setError] = useState(null);

    const navigate = useNavigate();

    // fetch questions from the backend
    useEffect(() => {
        const fetchQuestions = async () => {
            if (!currUser) {
                // navigate('/login');
                return;
            }

            try {
                const res = await Api.getQuestions();
                setQuestions(res);
            } catch (err) {
                setError(err.message)
            }
        };

        fetchQuestions();
    }, [currUser, navigate]);

    // handle answering a question
    const handleAnswer = (questionId, value) => {
        setAnswers({
            ...answers,
            [questionId]: value
        });
    };

    // go to the next question
    const handleNext = () => {
        if (currentQuestionIndex < questions.length - 1) {
            setCurrentQuestionIndex(currentQuestionIndex + 1);
        }
    };

    // go back to previous question
    const handlePrevious = () => {
        if (currentQuestionIndex > 0) {
            setCurrentQuestionIndex(currentQuestionIndex - 1);
        }
    };

    // submit quiz
    const handleSubmit = async () => {
        try {
            // submit answers
            const res = await Api.submitQuiz({ answers });

            // update the currUser context with the new MBTI result
            setCurrUser((user) => ({
                ...user,
                mbti: res.type,
            }));

            // navigate to results
            navigate('/results');
        } catch (err) {
            setError(err.message)
        }
    };

    if (error) return <p className="error">{error}</p>
    if (loading || questions === null) {
        return <Loading />
    }

    // track current question
    const currentQuestion = questions[currentQuestionIndex];

    // calculate progress as a percentage
    const progress = ((currentQuestionIndex + 1) / questions.length) * 100;

    return (
        <div className="QuizPage">
            {/* progress bar */}
            <div className="progress-bar-container">
                <div className="progress-bar" style={{ width: `${progress}%` }}></div>
            </div>

            {/* display current question */}
            <div className="question-container">
                <h4>{currentQuestion.text}</h4>
                <div className="options-container">
                    {currentQuestion.options.map((option) => (
                        // left and right options
                        <label key={option.value} className="option" >
                            <input
                                type="radio"
                                name={currentQuestion.id}
                                value={option.value}
                                onChange={() => handleAnswer(currentQuestion.id, option.value)}
                                checked={answers[currentQuestion.id] === option.value}
                                aria-label={option.text} // add for test accessibility
                            />
                            <span>{option.text}</span>
                        </label>
                    ))}
                </div>
            </div>

            {/* navigation buttons */}
            <div className="navigation-buttons">
                {currentQuestionIndex > 0 && (
                    <button
                        className="nav-button"
                        onClick={handlePrevious}
                        aria-label="Previous question">
                        <i className="fas fa-arrow-left fa-2x"></i>
                    </button>
                )}
                {currentQuestionIndex < questions.length - 1 ? (
                    <button
                        className="nav-button"
                        onClick={handleNext}
                        disabled={!answers[currentQuestion.id]}
                        aria-label="Next question">
                        <i className="fas fa-arrow-right fa-2x"></i>
                    </button>
                ) : (
                    answers[currentQuestion.id] && (
                        <button className="submit-button" onClick={handleSubmit}>
                            Submit
                        </button>
                    )
                )}
            </div>
        </div >
    );
};

export default QuizPage;
