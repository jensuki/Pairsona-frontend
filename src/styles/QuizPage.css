.QuizPage {
    max-width: var(--max-width-md);
}

.question-container h4 {
    margin-top: 50px;
    font-size: 1.2rem;
    font-weight: 500;
}

.options-container {
    margin-top: 50px;
    display: flex;
    justify-content: space-evenly;
}

.option {
    display: flex;
    align-items: center;
    border-radius: var(--border-radius-lg);
    cursor: pointer;
    transition: var(--transition-bg);
    background-color: #e0e0e0;
    margin-bottom: 25px;
}

/* choices */
.option span {
    font-weight: 500;
    padding: 13px;
}

/* selected choice */
input[type="radio"]:checked+span {
    color: var(--btn-white-text);
    background-color: var(--primary-green);
    padding: 13px;
    border-radius: var(--border-radius-lg);
}

.option input[type="radio"] {
    display: none;
}

.option:hover {
    background-color: var(--primary-green);
    color: white;
}

/* navigation buttons */
.navigation-buttons {
    margin-top: 20px;
    display: flex;
    justify-content: center;
    gap: 45px;
}

.nav-button {
    background: transparent;
    color: black;
    font-size: .6rem;

    &:hover {
        color: black;
        background: transparent;
    }
}

.submit-button {
    font-size: 1.1em;
    margin: 0;
    cursor: pointer;
    background-color: rgb(95, 82, 151);
    color: var(--btn-white-text);
    border: none;
    border-radius: var(--border-radius-sm);

    &:hover {
        background-color: rgb(75, 72, 147);
        transition: var(--transition-bg);
        animation: wiggle .4s ease-in;
    }
}

@keyframes wiggle {
    0% {
        transform: rotate(0deg);
    }

    25% {
        transform: rotate(-5deg);
    }

    50% {
        transform: rotate(3deg);
    }

    75% {
        transform: rotate(-2deg);
    }

    100% {
        transform: rotate(0);
    }
}

.progress-bar-container {
    width: 100%;
    height: 10px;
    background-color: #e0e0e0;
    border-radius: var(--border-radius-sm);
    overflow: hidden;
}

.progress-bar {
    height: 25px;
    background-color: var(--primary-green);
    transition: width 0.3s ease-in-out;
}