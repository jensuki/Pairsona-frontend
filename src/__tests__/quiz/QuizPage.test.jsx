import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../test_common/setupTests';
import QuizPage from '../../quiz/QuizPage';
import Api from '../../helpers/api';
import { describe, test, expect, vi, beforeEach } from 'vitest';
import { demoUser } from '../../test_common/mockData';

// mock api funcs + sample questions
vi.mock('../../helpers/api', () => ({
    default: {
        getQuestions: vi.fn().mockResolvedValue([
            {
                id: 1,
                text: 'Do you make lists or rely on memory?',
                options: [
                    { text: "makes lists", value: "S" },
                    { text: "relies on memory", value: "N" }
                ]
            },
            {
                id: 2,
                text: 'Do you prefer structured plans or spontaneity?',
                options: [
                    { text: "structured plans", value: "J" },
                    { text: "spontaneity", value: "P" }
                ]
            }
        ]),
        submitQuiz: vi.fn().mockResolvedValue({ type: "INFJ" }) // mock mbti set
    }
}))

// mock usenavigate
const mockNavigate = vi.fn();

vi.mock('react-router-dom', async (importActual) => {
    const actual = await importActual(); // import actual module
    return {
        ...actual, // keep original exports
        useNavigate: () => mockNavigate
    }
})

//helper
const renderQuizPage = () => {
    return render(
        <MemoryRouter>
            <UserProvider currUser={demoUser}>
                <QuizPage />
            </UserProvider>
        </MemoryRouter>
    );
};

describe('QuizPage component', () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    test('renders without crashing and displays first question', async () => {
        renderQuizPage();

        // wait for first question to load
        await waitFor(() => {
            expect(screen.getByText('Do you make lists or rely on memory?')).toBeInTheDocument();
        })

        expect(Api.getQuestions).toHaveBeenCalledTimes(1);
    })
    test('matches snapshot', async () => {
        const { asFragment } = renderQuizPage();

        // wait for the next arrow button to appear with class
        await waitFor(() => {
            const nextBtn = screen
                .getByRole('button', { name: /Next question/i })
                .querySelector('.fa-arrow-right');
            expect(nextBtn).toBeInTheDocument()
        });

        expect(asFragment()).toMatchSnapshot();
    });

    test('allows navigating between questions', async () => {
        const { getByText, getByRole } = renderQuizPage();

        await waitFor(() => {
            expect(getByText('Do you make lists or rely on memory?')).toBeInTheDocument();
        })

        // select first option
        const firstOption = getByRole('radio', { name: /makes lists/i })
        fireEvent.click(firstOption);
        // click next arrow
        const nextBtn = getByRole('button', { name: /next question/i });
        fireEvent.click(nextBtn);

        // wait or second question to appear
        await waitFor(() => {
            expect(getByText('Do you prefer structured plans or spontaneity?')).toBeInTheDocument();
        })

        // click back arrow button to go back to previous question
        const prevBtn = getByRole('button', { name: /previous question/i });
        fireEvent.click(prevBtn);

        // ensure first question is displayed again
        await waitFor(() => {
            expect(getByText('Do you make lists or rely on memory?')).toBeInTheDocument()
        })
    })

    test('disables next question button if no option is selected', async () => {
        const { getByText, getByRole } = renderQuizPage();

        await waitFor(() => {
            expect(getByText(/do you make lists or rely on memory?/i)).toBeInTheDocument();
        });

        // click next without choosing option
        const nextBtn = getByRole('button', { name: /next question/i });
        expect(nextBtn).toBeDisabled();
    })

    test('submits the quiz and redirects to /results page', async () => {
        const { getByRole } = renderQuizPage();

        // Wait for the first question to load
        await waitFor(() => {
            expect(screen.getByText(/do you make lists or rely on memory?/i)).toBeInTheDocument();
        });

        // answer all quiz questions
        const first = getByRole('radio', { name: /makes lists/i });
        fireEvent.click(first);

        const next = getByRole('button', { name: /next question/i });
        fireEvent.click(next)

        const second = getByRole('radio', { name: /structured plans/i });
        fireEvent.click(second);

        // click submit
        const submitBtn = getByRole('button', { name: /submit/i });
        fireEvent.click(submitBtn);

        // expect correct data
        expect(Api.submitQuiz).toHaveBeenCalledWith({
            answers: { 1: "S", 2: "J" }
        })

        // ensure redirection to /results
        await waitFor(() => {
            expect(mockNavigate).toHaveBeenCalledWith('/results');
        })
    })
    test('handles API errors gracefully', async () => {
        // mock an error msg to be thrown
        Api.getQuestions.mockRejectedValueOnce(new Error('Failed to fetch questions'));

        renderQuizPage();

        // wait for erro to appear
        await waitFor(() => {
            expect(screen.getByText(/failed to fetch questions/i)).toBeInTheDocument();
        });

        // no more spinner
        expect(screen.queryByRole('status')).not.toBeInTheDocument();

        // verify that no questions are displayed
        expect(screen.queryByText(/do you make lists or rely on memory?/i)).not.toBeInTheDocument();
    });
});
