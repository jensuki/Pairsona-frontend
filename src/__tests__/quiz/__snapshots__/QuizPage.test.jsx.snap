// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuizPage component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="QuizPage"
  >
    <div
      class="progress-bar-container"
    >
      <div
        class="progress-bar"
        style="width: 50%;"
      />
    </div>
    <div
      class="question-container"
    >
      <h4>
        Do you make lists or rely on memory?
      </h4>
      <div
        class="options-container"
      >
        <label
          class="option"
        >
          <input
            aria-label="makes lists"
            name="1"
            type="radio"
            value="S"
          />
          <span>
            makes lists
          </span>
        </label>
        <label
          class="option"
        >
          <input
            aria-label="relies on memory"
            name="1"
            type="radio"
            value="N"
          />
          <span>
            relies on memory
          </span>
        </label>
      </div>
    </div>
    <div
      class="navigation-buttons"
    >
      <button
        aria-label="Next question"
        class="nav-button"
        disabled=""
      >
        <i
          class="fas fa-arrow-right fa-2x"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
